### ISSUE TITLE
Actions implemented with global scope callbacks and not events?

### ISSUE THREAD

**Issue by rocketraman** [phonegap/phonegap-plugin-push#1192]
The PAYLOAD documentation describes actions implemented as callbacks on the global scope rather than as functions that can be registered directly on the plugin via something like `push.on('callbackName')` (similar to registration and notification). This seems inconsistent, and makes integration with frameworks such as Angular much more difficult.

Is there some fundamental reason why the implementation was done with callbacks expected on the global scope rather than registering the callback with the plugin?

**Comment by Birowsky** [comment]
Another obvious approach would be to let us specify action identifier instead of the globally scoped function.

**Comment by macdonst** [comment]
@rocketraman and @Birowsky it doesn't have to be a global scoped function but I do need it to be accessible in the global scope but it could be something like `app.pushservice.actions.accept` [Object->Object->Object->Function]. 

The reason, I coded it like that is so the plugin would know which function to call when the button was pressed. 

However, you have opened my eyes to a new way of doing this. Something like:

**Payload**

```
{
    "registration_ids": ["my device id"],
    "data": {
        "title": "AUX Scrum",
        "message": "Scrum: Daily touchbase @ 10am Please be on time so we can cover everything on the agenda.",
        "actions": [
            { "icon": "emailGuests", "title": "EMAIL GUESTS", "callback": "emailGuests", "foreground": true},
            { "icon": "snooze", "title": "SNOOZE", "callback": "snooze", "foreground": false}
        ]
    }
}
```

**Code**

```
push.on('emailGuests', function(data) {
    // email my guests
});

push.on('snooze', function(data) {
    // go to sleep
});
```

Basically register an event listener for the button press?

**Comment by rocketraman** [comment]
> Basically register an event listener for the button press?

Exactly.

**Comment by macdonst** [comment]
Okay, I think this is a really good idea and I will implement it in the future if someone else doesn't get to it first. I will set it up so the old way still works. Psuedo code:

```
// if a function named callback exists, execute it. Current behaviour.
if (callback) {
  callback();
} else {
  // use the call back name to look up what event listeners are registered for that name and execute each passing in the notification data.
  push.emit(callback, data);
}
```

**Comment by akz92** [comment]
I've forked the project and implemented events instead of callbacks. I didn't keep compatibility with the old way and didn't run the tests, but it's working fine. I plan on testing and then submitting a pull request. Any help is welcome :)

[My fork](https://github.com/akz92/phonegap-plugin-push)

**Comment by macdonst** [comment]
@akz92 the fork looks good, please submit a PR. However, I'd like where you do the `emit` to follow the pseudo-code I wrote above. First check to see there is an function by the name so folks who are used to the old API will still work.

**Comment by macdonst** [comment]
Fix released in 1.9.2
